File: java/jimmer-cloud/store-service/src/main/java/org/babyfish/jimmer/example/cloud/store/StoreService.java
package org.babyfish.jimmer.example.cloud.store;

import org.babyfish.jimmer.client.FetchBy;
import org.babyfish.jimmer.example.cloud.model.AuthorFetcher;
import org.babyfish.jimmer.example.cloud.model.BookFetcher;
import org.babyfish.jimmer.example.cloud.model.BookStore;
import org.babyfish.jimmer.example.cloud.model.BookStoreFetcher;
import org.babyfish.jimmer.sql.fetcher.Fetcher;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RestController;

@RestController
public class StoreService {

    private final StoreRepository storeRepository;

    public StoreService(StoreRepository storeRepository) {
        this.storeRepository = storeRepository;
    }

    /**
     * Load BookStore form current service
     * @param id The id of BookStore
     * @return The BookStore Object
     */
    @GetMapping("/store/{id}")
    public @FetchBy("SIMPLE_FETCHER") BookStore findStore(
            @PathVariable("id") long id
    ) {
        return storeRepository.findNullable(id, SIMPLE_FETCHER);
    }

    /**
     * Load BookStore form current service and associated objects from other services
     * @param id The id of BookStore
     * @return The BookStore object and associated objects
     */
    @GetMapping("/store/{id}/detail")
    public @FetchBy("COMPLEX_FETCHER") BookStore findStoreDetail(
            @PathVariable("id") long id
    ) {
        return storeRepository.findNullable(id, COMPLEX_FETCHER);
    }

    public static final Fetcher<BookStore> SIMPLE_FETCHER =
            BookStoreFetcher.$
                    .name();

    public static final Fetcher<BookStore> COMPLEX_FETCHER =
            BookStoreFetcher.$
                    .allScalarFields()
                    .books(
                            BookFetcher.$
                                    .allScalarFields()
                                    .authors(
                                            AuthorFetcher.$
                                                    .allScalarFields()
                                    )
                    );
}

File: kotlin/jimmer-cloud-kt/store-service/src/main/kotlin/org/babyfish/jimmer/example/cloud/kt/store/StoreService.kt
package org.babyfish.jimmer.example.cloud.kt.store

import org.babyfish.jimmer.client.FetchBy
import org.babyfish.jimmer.example.cloud.kt.model.BookStore
import org.babyfish.jimmer.example.cloud.kt.model.by
import org.babyfish.jimmer.sql.kt.fetcher.newFetcher
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RestController

@RestController
@Transactional
class StoreService(
    private val storeRepository: StoreRepository
) {

    /**
     * Load BookStore form current service
     * @param id The id of BookStore
     * @return The BookStore object
     */
    @GetMapping("/store/{id}")
    fun findStore(
        @PathVariable("id") id: Long
    ): @FetchBy("SIMPLE_FETCHER") BookStore? =
        storeRepository.findNullable(id, SIMPLE_FETCHER)

    /**
     * Load BookStore form current service and associated objects from other services
     * @param id The id of BookStore
     * @return The BookStore object and associated objects
     */
    @GetMapping("/store/{id}/detail")
    fun findStoreDetail(
        @PathVariable("id") id: Long
    ): @FetchBy("COMPLEX_FETCHER") BookStore? =
        storeRepository.findNullable(id, COMPLEX_FETCHER)

    companion object {

        val SIMPLE_FETCHER = newFetcher(BookStore::class).by {
            name()
        }

        val COMPLEX_FETCHER = newFetcher(BookStore::class).by {
            allScalarFields()
            books {
                allScalarFields()
                authors {
                    allScalarFields()
                }
            }
        }
    }
}